File: src/storage/repositories/activity.repository.ts
/**
 * 活动数据访问层
 */

import Database from 'better-sqlite3';
import { Activity, ActivityQueryOptions, ActivityStatistics, ActivityType } from '../../domain/activity';

export class ActivityRepository {
  constructor(private db: Database.Database) {}

  /**
   * 插入活动记录
   */
  insert(activity: Activity): number {
    const stmt = this.db.prepare(`
      INSERT INTO activities (type, timestamp, duration, data)
      VALUES (?, ?, ?, ?)
    `);

    const result = stmt.run(
      activity.type,
      activity.timestamp,
      activity.duration || 0,
      JSON.stringify(activity.data)
    );

    return result.lastInsertRowid as number;
  }

  /**
   * 批量插入活动记录
   */
  insertBatch(activities: Activity[]): number {
    const stmt = this.db.prepare(`
      INSERT INTO activities (type, timestamp, duration, data)
      VALUES (?, ?, ?, ?)
    `);

    const transaction = this.db.transaction((acts: Activity[]) => {
      for (const activity of acts) {
        stmt.run(
          activity.type,
          activity.timestamp,
          activity.duration || 0,
          JSON.stringify(activity.data)
        );
      }
    });

    transaction(activities);
    return activities.length;
  }

  /**
   * 根据ID查询活动
   */
  findById(id: number): Activity | null {
    const stmt = this.db.prepare('SELECT * FROM activities WHERE id = ?');
    const row = stmt.get(id) as any;

    if (!row) {
      return null;
    }

    return this.mapRowToActivity(row);
  }

  /**
   * 查询活动列表
   */
  query(options: ActivityQueryOptions = {}): Activity[] {
    let sql = 'SELECT * FROM activities WHERE 1=1';
    const params: any[] = [];

    if (options.startTime !== undefined) {
      sql += ' AND timestamp >= ?';
      params.push(options.startTime);
    }

    if (options.endTime !== undefined) {
      sql += ' AND timestamp <= ?';
      params.push(options.endTime);
    }

    if (options.type !== undefined) {
      sql += ' AND type = ?';
      params.push(options.type);
    }

    sql += ' ORDER BY timestamp DESC';

    if (options.limit !== undefined) {
      sql += ' LIMIT ?';
      params.push(options.limit);
    }

    if (options.offset !== undefined) {
      sql += ' OFFSET ?';
      params.push(options.offset);
    }

    const stmt = this.db.prepare(sql);
    const rows = stmt.all(...params) as any[];

    return rows.map(row => this.mapRowToActivity(row));
  }

  /**
   * 查询最近的活动
   */
  findRecent(limit = 10): Activity[] {
    return this.query({ limit });
  }

  /**
   * 查询指定时间范围内的活动
   */
  findByTimeRange(startTime: number, endTime: number): Activity[] {
    return this.query({ startTime, endTime });
  }

  /**
   * 删除指定时间之前的活动
   */
  deleteBefore(timestamp: number): number {
    const stmt = this.db.prepare('DELETE FROM activities WHERE timestamp < ?');
    const result = stmt.run(timestamp);
    return result.changes;
  }

  /**
   * 清空所有活动记录
   */
  deleteAll(): number {
    const stmt = this.db.prepare('DELETE FROM activities');
    const result = stmt.run();
    return result.changes;
  }

  /**
   * 获取活动统计信息
   */
  getStatistics(startTime?: number, endTime?: number): ActivityStatistics {
    let whereClause = 'WHERE 1=1';
    const params: any[] = [];

    if (startTime !== undefined) {
      whereClause += ' AND timestamp >= ?';
      params.push(startTime);
    }

    if (endTime !== undefined) {
      whereClause += ' AND timestamp <= ?';
      params.push(endTime);
    }

    // 总时长
    const totalDurationStmt = this.db.prepare(
      `SELECT SUM(duration) as total FROM activities ${whereClause}`
    );
    const { total } = totalDurationStmt.get(...params) as any;

    // 各类型时长
    const byTypeStmt = this.db.prepare(`
      SELECT type, SUM(duration) as duration
      FROM activities ${whereClause}
      GROUP BY type
    `);
    const byTypeRows = byTypeStmt.all(...params) as any[];
    const durationByType: Record<ActivityType, number> = {
      [ActivityType.BROWSER]: 0,
      [ActivityType.APPLICATION]: 0,
      [ActivityType.SYSTEM]: 0
    };

    for (const row of byTypeRows) {
      durationByType[row.type as ActivityType] = row.duration || 0;
    }

    // 最频繁的活动（前10）
    const topActivitiesStmt = this.db.prepare(`
      SELECT
        json_extract(data, '$.url') as name,
        SUM(duration) as duration,
        COUNT(*) as count
      FROM activities ${whereClause}
      WHERE type = 'browser'
      GROUP BY name
      ORDER BY duration DESC
      LIMIT 10
    `);
    const topActivities = topActivitiesStmt.all(...params) as any[];

    return {
      totalDuration: total || 0,
      durationByType,
      topActivities: topActivities.map(item => ({
        name: item.name || 'Unknown',
        duration: item.duration,
        count: item.count
      })),
      distractionRatio: 0 // 将在judgment repository中计算
    };
  }

  /**
   * 将数据库行映射为Activity对象
   */
  private mapRowToActivity(row: any): Activity {
    return {
      id: row.id,
      type: row.type,
      timestamp: row.timestamp,
      duration: row.duration,
      data: JSON.parse(row.data)
    };
  }
}

File: python/storage/activity_repository.py
"""
活动数据仓库

作用：
1. 管理活动记录的CRUD操作
2. 提供查询统计功能
"""

import json
from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta
from .database import db_manager


class ActivityRepository:
    """活动数据仓库"""

    def __init__(self):
        self.db = db_manager

    def save_activity(
        self,
        activity_type: str,
        timestamp: int,
        duration: int,
        data: Dict[str, Any]
    ) -> int:
        """
        保存活动记录

        Args:
            activity_type: 活动类型 (application/browser)
            timestamp: 时间戳（毫秒）
            duration: 持续时间（毫秒）
            data: 活动数据（字典）

        Returns:
            新记录的ID
        """
        conn = self.db.get_connection()

        cursor = conn.execute('''
            INSERT INTO activities (type, timestamp, duration, data)
            VALUES (?, ?, ?, ?)
        ''', (
            activity_type,
            timestamp,
            duration,
            json.dumps(data, ensure_ascii=False)
        ))

        conn.commit()
        return cursor.lastrowid

    def get_recent_activities(
        self,
        limit: int = 100,
        activity_type: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        获取最近的活动记录

        Args:
            limit: 返回数量限制
            activity_type: 活动类型过滤（可选）

        Returns:
            活动记录列表
        """
        conn = self.db.get_connection()

        if activity_type:
            cursor = conn.execute('''
                SELECT * FROM activities
                WHERE type = ?
                ORDER BY timestamp DESC
                LIMIT ?
            ''', (activity_type, limit))
        else:
            cursor = conn.execute('''
                SELECT * FROM activities
                ORDER BY timestamp DESC
                LIMIT ?
            ''', (limit,))

        rows = cursor.fetchall()
        return [dict(row) for row in rows]

    def get_activities_by_timerange(
        self,
        start_time: int,
        end_time: int,
        activity_type: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        获取指定时间范围内的活动记录

        Args:
            start_time: 开始时间戳（毫秒）
            end_time: 结束时间戳（毫秒）
            activity_type: 活动类型过滤（可选）

        Returns:
            活动记录列表
        """
        conn = self.db.get_connection()

        if activity_type:
            cursor = conn.execute('''
                SELECT * FROM activities
                WHERE type = ? AND timestamp >= ? AND timestamp <= ?
                ORDER BY timestamp DESC
            ''', (activity_type, start_time, end_time))
        else:
            cursor = conn.execute('''
                SELECT * FROM activities
                WHERE timestamp >= ? AND timestamp <= ?
                ORDER BY timestamp DESC
            ''', (start_time, end_time))

        rows = cursor.fetchall()
        return [dict(row) for row in rows]

    def get_total_duration(
        self,
        start_time: Optional[int] = None,
        end_time: Optional[int] = None,
        activity_type: Optional[str] = None
    ) -> int:
        """
        获取总持续时间

        Args:
            start_time: 开始时间戳（毫秒）
            end_time: 结束时间戳（毫秒）
            activity_type: 活动类型过滤（可选）

        Returns:
            总持续时间（毫秒）
        """
        conn = self.db.get_connection()

        if start_time and end_time:
            if activity_type:
                cursor = conn.execute('''
                    SELECT SUM(duration) as total
                    FROM activities
                    WHERE type = ? AND timestamp >= ? AND timestamp <= ?
                ''', (activity_type, start_time, end_time))
            else:
                cursor = conn.execute('''
                    SELECT SUM(duration) as total
                    FROM activities
                    WHERE timestamp >= ? AND timestamp <= ?
                ''', (start_time, end_time))
        else:
            if activity_type:
                cursor = conn.execute('''
                    SELECT SUM(duration) as total
                    FROM activities
                    WHERE type = ?
                ''', (activity_type,))
            else:
                cursor = conn.execute('''
                    SELECT SUM(duration) as total
                    FROM activities
                ''')

        result = cursor.fetchone()
        return result['total'] or 0

    def get_activity_count(
        self,
        activity_type: Optional[str] = None
    ) -> int:
        """
        获取活动记录总数

        Args:
            activity_type: 活动类型过滤（可选）

        Returns:
            活动记录数量
        """
        conn = self.db.get_connection()

        if activity_type:
            cursor = conn.execute('''
                SELECT COUNT(*) as count
                FROM activities
                WHERE type = ?
            ''', (activity_type,))
        else:
            cursor = conn.execute('''
                SELECT COUNT(*) as count
                FROM activities
            ''')

        result = cursor.fetchone()
        return result['count']

    def get_top_apps(
        self,
        hours: int = 24,
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """
        获取使用时间最长的应用

        Args:
            hours: 统计最近多少小时
            limit: 返回数量限制

        Returns:
            应用使用统计列表
        """
        conn = self.db.get_connection()

        # 计算时间范围
        end_time = int(datetime.now().timestamp() * 1000)
        start_time = end_time - (hours * 3600 * 1000)

        cursor = conn.execute('''
            SELECT
                json_extract(data, '$.app_name') as app_name,
                SUM(duration) as total_duration,
                COUNT(*) as activity_count
            FROM activities
            WHERE type = 'application'
            AND timestamp >= ? AND timestamp <= ?
            GROUP BY app_name
            ORDER BY total_duration DESC
            LIMIT ?
        ''', (start_time, end_time, limit))

        rows = cursor.fetchall()
        return [dict(row) for row in rows]

    def delete_old_activities(self, days: int = 30):
        """
        删除旧的活动记录

        Args:
            days: 保留最近多少天的数据
        """
        conn = self.db.get_connection()

        cutoff_time = int((datetime.now() - timedelta(days=days)).timestamp() * 1000)

        conn.execute('''
            DELETE FROM activities
            WHERE timestamp < ?
        ''', (cutoff_time,))

        conn.commit()
        print(f'已删除 {days} 天前的活动记录')


# 全局活动仓库实例
activity_repository = ActivityRepository()
